test/functional/options/defaults_spec.lua

```diff
@@ -328,3 +332,235 @@ describe('XDG-based defaults', function()
+        alter_path('/etc/config/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it('handles changes during runtime', function()
+      clear({env={
+        XDG_CONFIG_DIRS=alter_path('/home/original:/etc/config')
+      }})
+      eq({
+        alter_path('/home/original/nvim'),
+        alter_path('/etc/config/nvim')
+      }, funcs.stdpath('config_dirs'))
+      nvim('command', 'let $XDG_CONFIG_DIRS="/home/new:/etc/config"')
+      eq({
+        alter_path('/home/new/nvim'),
+        alter_path('/etc/config/nvim')
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it("doesn't expand $VAR and $IBLES", function()
+      clear({env={
+        XDG_CONFIG_DIRS='$VAR:$IBLES',
+        VAR='this-should-not-happen',
+        IBLES='neither-should-this'
+      }})
+      eq({
+        alter_path('$VAR/nvim'),
+        alter_path('$IBLES/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_CONFIG_DIRS='~/.oldconfig:~/.olderconfig'
+      }})
+      eq({
+        alter_path('~/.oldconfig/nvim'),
+        alter_path('~/.olderconfig/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+  end)
+
+  describe('"data_dirs"' , function ()
+    it('knows XDG_DATA_DIRS', function()
+      clear({env={
+        XDG_DATA_DIRS=alter_path('/home/docwhat/.data:/etc/local')
+      }})
+      eq({
+        alter_path('/home/docwhat/.data/nvim'),
+        alter_path('/etc/local/nvim'),
+      }, funcs.stdpath('data_dirs'))
+    end)
+
+    it('handles changes during runtime', function()
+      clear({env={
+        XDG_DATA_DIRS=alter_path('/home/original:/etc/local')
+      }})
+      eq({
+        alter_path('/home/original/nvim'),
+        alter_path('/etc/local/nvim')
+      }, funcs.stdpath('data_dirs'))
+      nvim('command', 'let $XDG_DATA_DIRS="/home/new:/etc/local"')
+      eq({
+        alter_path('/home/new/nvim'),
+        alter_path('/etc/local/nvim')
+      }, funcs.stdpath('data_dirs'))
+    end)
+
+    it("doesn't expand $VAR and $IBLES", function()
+      clear({env={
+        XDG_DATA_DIRS='$VAR:$IBLES',
+        VAR='this-should-not-happen',
+        IBLES='neither-should-this'
+      }})
+      eq({
+        alter_path('$VAR/nvim'),
+        alter_path('$IBLES/nvim'),
+      }, funcs.stdpath('data_dirs'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_DATA_DIRS='~/.olddata:~/.olderdata'
+      }})
+      eq({
+        alter_path('~/.olddata/nvim'),
+        alter_path('~/.olderdata/nvim'),
+      }, funcs.stdpath('data_dirs'))
+    end)
+  end)
+
+  it('errors on unknown strings', function()
+    eq('Vim(call):E474: Invalid argument', exc_exec('call stdpath("capybara")'))
+    eq('Vim(call):E474: Invalid argument', exc_exec('call stdpath("")'))
+  end)
+
+  it('errors on non-strings', function()
+    eq('Vim(call):E474: Invalid argument', exc_exec('call stdpath(23)'))
```

Comment:
Numbers are coerced to strings, so it really goes to “errors on unknown strings”.